build proxy rpm failed

**Bug description**
tried to build sidecar rpm for RedHat VM mesh expansion, but failed with below error:

```
FAILED: Build did NOT complete successfully
error: Bad exit status from /var/tmp/rpm-tmp.bkTcOx (%build)
    Bad exit status from /var/tmp/rpm-tmp.bkTcOx (%build)
[builderr_log.txt](https://github.com/istio/istio/files/3375431/builderr_log.txt)
[builderr_log.txt](https://github.com/istio/istio/files/3375434/builderr_log.txt)

RPM build errors:
Could not execute local: rpmbuild --define '_sourcedir /tmp/tmp.HGP34eOTBQ/packaging/rpm/proxy' --define '_specdir /tmp/tmp.HGP34eOTBQ/packaging/rpm/proxy' --define '_builddir /tmp/tmp.HGP34eOTBQ/packaging/rpm/proxy' --define '_srcrpmdir /tmp/tmp.HGP34eOTBQ/packaging/rpm/proxy' --define '_rpmdir /tmp/tmp.HGP34eOTBQ/packaging/rpm/proxy' --define 'dist %{?distprefix}.el7' --define 'rhel 7' --eval '%  undefine fedora' --define 'el7 1' -ba /tmp/tmp.HGP34eOTBQ/packaging/rpm/proxy/istio-proxy.spec | tee .build-1.1.0-1.el7.log
[builderr_log.txt](https://github.com/istio/istio/files/3375425/builderr_log.txt)

```

**Affected product area (please put an X in all that apply)**

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Expected behavior**

**Steps to reproduce the bug**
Steps to build the RPM package:
1. Clone Istio repo (https://github.com/istio/istio.git) into dir /root/git/src/istio and cd into this dir
2. Run cmd make VERSION=1.2.0 rpm/builder-image to build docker builder image: istio-rpm-builder
3. Run cmd make VERSION=1.2.0 rpm/proxy to build, but had some git issue during build
4. Try below command to debug the issue:
docker run -it \
        -v ${GO_TOP}:${GO_TOP} \
				-w /builder \
        -e USER=${USER} \
				-e ISTIO_ENVOY_VERSION=${ISTIO_ENVOY_VERSION} \
				-e ISTIO_GO=${ISTIO_GO} \
				-e ISTIO_OUT=${ISTIO_OUT} \
				-e PACKAGE_VERSION=${PACKAGE_VERSION} \
				-e USER_ID=$(shell id -u) \
				-e GROUP_ID=$(shell id -g) \
				istio-rpm-builder \
				bash
5. In docker constainer bash, run script in tools/packaging/rpm/build-proxy-rpm.sh, found the issue was because CentOS's Git version is too low, so upgrade Git to latest version
6. Run tools/packaging/rpm/build-proxy-rpm.sh again, got C++ compilation errors, logs are attached.

**Version (include the output of `istioctl version --remote` and `kubectl version`)**

**How was Istio installed?**

**Environment where bug was observed (cloud vendor, OS, etc)**
RedHat

Additionally, please consider attaching a [cluster state archive](http://istio.io/help/bugs/#generating-a-cluster-state-archive) by attaching
the dump file to this issue.

INFO: From Generating Descriptor Set proto_library @envoy_api//envoy/service/accesslog/v2:als:
external/com_google_protobuf: warning: directory does not exist.
INFO: From Generating Descriptor Set proto_library @envoy_api//envoy/config/cluster/redis:redis_cluster:
external/com_google_protobuf: warning: directory does not exist.
INFO: From Generating Descriptor Set proto_library @envoy_api//envoy/config/transport_socket/alts/v2alpha:alts:
external/com_google_protobuf: warning: directory does not exist.
INFO: From CcConfigureMakeRule external/envoy/bazel/foreign_cc/luajit/include:

INFO: From Generating Descriptor Set proto_library @envoy_api//envoy/service/discovery/v2:sds:
external/com_google_protobuf: warning: directory does not exist.
INFO: From Generating Descriptor Set proto_library @envoy_api//envoy/service/discovery/v2:ads:
external/com_google_protobuf: warning: directory does not exist.
INFO: From CcConfigureMakeRule external/envoy/bazel/foreign_cc/gperftools_build/include:

ERROR: /builder/bazel_cache/external/envoy/source/extensions/filters/network/kafka/BUILD:92:1: C++ compilation of rule '@envoy//source/extensions/filters/network/kafka:serialization_lib' failed (Exit 1) envoy_cc_wrapper failed: error executing command /builder/bazel_cache/external/local_config_cc/extra_tools/envoy_cc_wrapper -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer ... (remaining 94 argument(s) skipped)

Use --sandbox_debug to see verbose messages from the sandbox
In file included from external/envoy/source/extensions/filters/network/kafka/serialization.cc:1:0:
bazel-out/k8-fastbuild/bin/external/envoy/source/extensions/filters/network/kafka/_virtual_includes/serialization_lib/extensions/filters/network/kafka/serialization.h: In member function 'uint32_t Envoy::Extensions::NetworkFilters::Kafka::EncodingContext::computeSize(const T&) const [with T = bool; uint32_t = unsigned int]':
bazel-out/k8-fastbuild/bin/external/envoy/source/extensions/filters/network/kafka/_virtual_includes/serialization_lib/extensions/filters/network/kafka/serialization.h:526:34: error: enclosing class of constexpr non-static member function 'uint32_t Envoy::Extensions::NetworkFilters::Kafka::EncodingContext::computeSize(const T&) const [with T = bool; uint32_t = unsigned int]' is not a literal type
   template <> constexpr uint32_t EncodingContext::computeSize(const TYPE&) const {                 \
                                  ^
bazel-out/k8-fastbuild/bin/external/envoy/source/extensions/filters/network/kafka/_virtual_includes/serialization_lib/extensions/filters/network/kafka/serialization.h:530:1: note: in expansion of macro 'COMPUTE_SIZE_OF_NUMERIC_TYPE'
 COMPUTE_SIZE_OF_NUMERIC_TYPE(bool)
 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
bazel-out/k8-fastbuild/bin/external/envoy/source/extensions/filters/network/kafka/_virtual_includes/serialization_lib/extensions/filters/network/kafka/serialization.h:468:7: note: 'Envoy::Extensions::NetworkFilters::Kafka::EncodingContext' is not literal because:
 class EncodingContext {
       ^~~~~~~~~~~~~~~
bazel-out/k8-fastbuild/bin/external/envoy/source/extensions/filters/network/kafka/_virtual_includes/serialization_lib/extensions/filters/network/kafka/serialization.h:468:7: note:   'Envoy::Extensions::NetworkFilters::Kafka::EncodingContext' is not an aggregate, does not have a trivial default constructor, and has no constexpr constructor that is not a copy or move constructor
bazel-out/k8-fastbuild/bin/external/envoy/source/extensions/filters/network/kafka/_virtual_includes/serialization_lib/extensions/filters/network/kafka/serialization.h: In member function 'uint32_t Envoy::Extensions::NetworkFilters::Kafka::EncodingContext::computeSize(const T&) const [with T = signed char; uint32_t = unsigned int]':
bazel-out/k8-fastbuild/bin/external/envoy/source/extensions/filters/network/kafka/_virtual_includes/serialization_lib/extensions/filters/network/kafka/serialization.h:526:34: error: enclosing class of constexpr non-static member function 'uint32_t Envoy::Extensions::NetworkFilters::Kafka::EncodingContext::computeSize(const T&) const [with T = signed char; uint32_t = unsigned int]' is not a literal type
   template <> constexpr uint32_t EncodingContext::computeSize(const TYPE&) const {                 \
                                  ^
bazel-out/k8-fastbuild/bin/external/envoy/source/extensions/filters/network/kafka/_virtual_includes/serialization_lib/extensions/filters/network/kafka/serialization.h:531:1: note: in expansion of macro 'COMPUTE_SIZE_OF_NUMERIC_TYPE'
 COMPUTE_SIZE_OF_NUMERIC_TYPE(int8_t)
 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
bazel-out/k8-fastbuild/bin/external/envoy/source/extensions/filters/network/kafka/_virtual_includes/serialization_lib/extensions/filters/network/kafka/serialization.h: In member function 'uint32_t Envoy::Extensions::NetworkFilters::Kafka::EncodingContext::computeSize(const T&) const [with T = short int; uint32_t = unsigned int]':
bazel-out/k8-fastbuild/bin/external/envoy/source/extensions/filters/network/kafka/_virtual_includes/serialization_lib/extensions/filters/network/kafka/serialization.h:526:34: error: enclosing class of constexpr non-static member function 'uint32_t Envoy::Extensions::NetworkFilters::Kafka::EncodingContext::computeSize(const T&) const [with T = short int; uint32_t = unsigned int]' is not a literal type
   template <> constexpr uint32_t EncodingContext::computeSize(const TYPE&) const {                 \
                                  ^
bazel-out/k8-fastbuild/bin/external/envoy/source/extensions/filters/network/kafka/_virtual_includes/serialization_lib/extensions/filters/network/kafka/serialization.h:532:1: note: in expansion of macro 'COMPUTE_SIZE_OF_NUMERIC_TYPE'
 COMPUTE_SIZE_OF_NUMERIC_TYPE(int16_t)
 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
bazel-out/k8-fastbuild/bin/external/envoy/source/extensions/filters/network/kafka/_virtual_includes/serialization_lib/extensions/filters/network/kafka/serialization.h: In member function 'uint32_t Envoy::Extensions::NetworkFilters::Kafka::EncodingContext::computeSize(const T&) const [with T = int; uint32_t = unsigned int]':
bazel-out/k8-fastbuild/bin/external/envoy/source/extensions/filters/network/kafka/_virtual_includes/serialization_lib/extensions/filters/network/kafka/serialization.h:526:34: error: enclosing class of constexpr non-static member function 'uint32_t Envoy::Extensions::NetworkFilters::Kafka::EncodingContext::computeSize(const T&) const [with T = int; uint32_t = unsigned int]' is not a literal type
   template <> constexpr uint32_t EncodingContext::computeSize(const TYPE&) const {                 \
                                  ^
bazel-out/k8-fastbuild/bin/external/envoy/source/extensions/filters/network/kafka/_virtual_includes/serialization_lib/extensions/filters/network/kafka/serialization.h:533:1: note: in expansion of macro 'COMPUTE_SIZE_OF_NUMERIC_TYPE'
 COMPUTE_SIZE_OF_NUMERIC_TYPE(int32_t)
 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
bazel-out/k8-fastbuild/bin/external/envoy/source/extensions/filters/network/kafka/_virtual_includes/serialization_lib/extensions/filters/network/kafka/serialization.h: In member function 'uint32_t Envoy::Extensions::NetworkFilters::Kafka::EncodingContext::computeSize(const T&) const [with T = unsigned int; uint32_t = unsigned int]':
bazel-out/k8-fastbuild/bin/external/envoy/source/extensions/filters/network/kafka/_virtual_includes/serialization_lib/extensions/filters/network/kafka/serialization.h:526:34: error: enclosing class of constexpr non-static member function 'uint32_t Envoy::Extensions::NetworkFilters::Kafka::EncodingContext::computeSize(const T&) const [with T = unsigned int; uint32_t = unsigned int]' is not a literal type
   template <> constexpr uint32_t EncodingContext::computeSize(const TYPE&) const {                 \
                                  ^
bazel-out/k8-fastbuild/bin/external/envoy/source/extensions/filters/network/kafka/_virtual_includes/serialization_lib/extensions/filters/network/kafka/serialization.h:534:1: note: in expansion of macro 'COMPUTE_SIZE_OF_NUMERIC_TYPE'
 COMPUTE_SIZE_OF_NUMERIC_TYPE(uint32_t)
 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
bazel-out/k8-fastbuild/bin/external/envoy/source/extensions/filters/network/kafka/_virtual_includes/serialization_lib/extensions/filters/network/kafka/serialization.h: In member function 'uint32_t Envoy::Extensions::NetworkFilters::Kafka::EncodingContext::computeSize(const T&) const [with T = long int; uint32_t = unsigned int]':
bazel-out/k8-fastbuild/bin/external/envoy/source/extensions/filters/network/kafka/_virtual_includes/serialization_lib/extensions/filters/network/kafka/serialization.h:526:34: error: enclosing class of constexpr non-static member function 'uint32_t Envoy::Extensions::NetworkFilters::Kafka::EncodingContext::computeSize(const T&) const [with T = long int; uint32_t = unsigned int]' is not a literal type
   template <> constexpr uint32_t EncodingContext::computeSize(const TYPE&) const {                 \
                                  ^
bazel-out/k8-fastbuild/bin/external/envoy/source/extensions/filters/network/kafka/_virtual_includes/serialization_lib/extensions/filters/network/kafka/serialization.h:535:1: note: in expansion of macro 'COMPUTE_SIZE_OF_NUMERIC_TYPE'
 COMPUTE_SIZE_OF_NUMERIC_TYPE(int64_t)
 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
INFO: Elapsed time: 595.475s, Critical Path: 42.98s
INFO: 832 processes: 832 processwrapper-sandbox.
FAILED: Build did NOT complete successfully
error: Bad exit status from /var/tmp/rpm-tmp.bkTcOx (%build)
    Bad exit status from /var/tmp/rpm-tmp.bkTcOx (%build)


RPM build errors:
Could not execute local: rpmbuild --define '_sourcedir /tmp/tmp.HGP34eOTBQ/packaging/rpm/proxy' --define '_specdir /tmp/tmp.HGP34eOTBQ/packaging/rpm/proxy' --define '_builddir /tmp/tmp.HGP34eOTBQ/packaging/rpm/proxy' --define '_srcrpmdir /tmp/tmp.HGP34eOTBQ/packaging/rpm/proxy' --define '_rpmdir /tmp/tmp.HGP34eOTBQ/packaging/rpm/proxy' --define 'dist %{?distprefix}.el7' --define 'rhel 7' --eval '%undefine fedora' --define 'el7 1' -ba /tmp/tmp.HGP34eOTBQ/packaging/rpm/proxy/istio-proxy.spec | tee .build-1.1.0-1.el7.log

--- a/tools/packaging/rpm/Dockerfile.build
+++ b/tools/packaging/rpm/Dockerfile.build
@@ -2,13 +2,19 @@ FROM centos:7
 
 RUN yum upgrade -y && \
     yum install -y epel-release centos-release-scl && \
-    yum install -y fedpkg golang sudo make which cmake3 \
-                   automake autoconf autogen libtool \
-                   devtoolset-6-gcc devtoolset-6-gcc-c++ \
-                   devtoolset-6-libatomic-devel ninja-build && \
+    yum install -y fedpkg sudo make which cmake3 rh-git218 \
+                   automake autoconf autogen libtool ninja-build \
+                   llvm-toolset-7 devtoolset-7-libatomic-devel && \
     yum clean all
 
-RUN curl -o /root/bazel-installer.sh -L http://github.com/bazelbuild/bazel/releases/download/0.22.0/bazel-0.22.0-installer-linux-x86_64.sh && \
+# Install go (go package in yum repositories is too old)
+RUN curl -o /root/go.tar.gz https://dl.google.com/go/go1.12.7.linux-amd64.tar.gz && \
+    tar zxf /root/go.tar.gz -C /usr/local
+ENV GOROOT=/usr/local/go \
+    PATH=/usr/local/go/bin:${PATH}
+
+# Install bazel
+RUN curl -o /root/bazel-installer.sh -L http://github.com/bazelbuild/bazel/releases/download/0.27.1/bazel-0.27.1-installer-linux-x86_64.sh && \
     chmod +x /root/bazel-installer.sh && \
     /root/bazel-installer.sh
 

--- a/tools/packaging/rpm/build-proxy-rpm.sh
+++ b/tools/packaging/rpm/build-proxy-rpm.sh
@@ -17,7 +17,9 @@ cp -r "${PKG_DIR}" "${WORK_DIR}"
 mv "${WORK_DIR}"/packaging/common/* "${RPM_DIR}"
 
 # shellcheck disable=SC1091
-source /opt/rh/devtoolset-6/enable
+source /opt/rh/rh-git218/enable
+# shellcheck disable=SC1091
+source /opt/rh/llvm-toolset-7/enable
 
 cd /builder || exit 1
 git clone  https://github.com/istio/proxy.git istio-proxy

--- a/tools/packaging/rpm/istio/istio.spec
+++ b/tools/packaging/rpm/istio/istio.spec
@@ -44,8 +44,6 @@ Source7:        istio-auth-node-agent.service
 
 # e.g. el6 has ppc64 arch without gcc-go, so EA tag is required
 ExclusiveArch:  %{?go_arches:%{go_arches}}%{!?go_arches:%{ix86} x86_64 aarch64 %{arm}}
-# If go_compiler is not set to 1, there is no virtual provide. Use golang instead.
-BuildRequires:  golang >= 1.12
 
 %description
 Istio is an open platform that provides a uniform way to connect, manage
@@ -217,7 +215,7 @@ tar zxf %{SOURCE0} -C ISTIO/src/istio.io/istio --strip=1
 
 %build
 cd ISTIO
-export GOPATH=$(pwd):%{gopath}
+export GOPATH=$(pwd)
 
 pushd src/istio.io/istio
 make pilot-discovery pilot-agent istioctl sidecar-injector mixc mixs citadel galley node_agent

--- a/tools/packaging/rpm/proxy/istio-proxy.spec
+++ b/tools/packaging/rpm/proxy/istio-proxy.spec
@@ -21,12 +21,6 @@ License:        ASL 2.0
 URL:            https://%{provider_prefix}
 
 BuildRequires:  ninja-build
-BuildRequires:  devtoolset-6-gcc
-BuildRequires:  devtoolset-6-gcc-c++
-BuildRequires:  devtoolset-6-libatomic-devel
-BuildRequires:  devtoolset-6-libstdc++-devel
-BuildRequires:  devtoolset-6-runtime
-BuildRequires:  golang
 BuildRequires:  perl
 BuildRequires:  binutils
 BuildRequires:  cmake3
@@ -54,6 +48,8 @@ istio-proxy is the proxy required by the Istio Pilot Agent that talks to Istio p
 %setup -q -n %{name}
 
 %build
+export CC=clang
+export CXX=clang++
 bazel --output_base=/builder/bazel_cache --output_user_root=/builder/bazel_cache/root  build //...
 bazel shutdown
 

